rl4j/rl4j-core/src/test/java/org/deeplearning4j/rl4j/TMazeExample.java

```diff
@@ -0,0 +1,247 @@
+                .agentLearnerListeners(listeners)
+                .asynchronous(isAsync)
+                .build();
+        return new NStepQLearningBuilder(configuration, network, environmentBuilder, transformProcessBuilder, rnd);
+    }
+
+    private static Builder<IAgentLearner<Integer>> setupAdvantageActorCritic(Builder<Environment<Integer>> environmentBuilder,
+                                                                             Builder<TransformProcess> transformProcessBuilder,
+                                                                             List<AgentListener<Integer>> listeners,
+                                                                             Random rnd, boolean isAsync) {
+        ITrainableNeuralNet<ActorCriticCompGraph> network = buildActorCriticNetwork();
+
+        AdvantageActorCriticBuilder.Configuration configuration = AdvantageActorCriticBuilder.Configuration.builder()
+                .neuralNetUpdaterConfiguration(NeuralNetUpdaterConfiguration.builder()
+                        .build())
+                .advantageActorCriticConfiguration(AdvantageActorCritic.Configuration.builder()
+                        .gamma(0.99)
+                        .build())
+                .experienceHandlerConfiguration(StateActionExperienceHandler.Configuration.builder()
+                        .batchSize(Integer.MAX_VALUE)
+                        .build())
+                .agentLearnerConfiguration(AgentLearner.Configuration.builder()
+                        .maxEpisodeSteps(40)
+                        .build())
+                .agentLearnerListeners(listeners)
+                .asynchronous(isAsync)
+                .build();
+        return new AdvantageActorCriticBuilder(configuration, network, environmentBuilder, transformProcessBuilder, rnd);
+    }
+
+    private static ComputationGraphConfiguration.GraphBuilder buildBaseNetworkConfiguration() {
+        return new NeuralNetConfiguration.Builder().seed(Constants.NEURAL_NET_SEED)
+                        .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
+                        .updater(new Adam())
+                        .weightInit(WeightInit.XAVIER)
+                        .graphBuilder()
+                        .setInputTypes(InputType.recurrent(NUM_INPUTS))
+                        .addInputs("input")
+                        .addLayer("goal", new LSTM.Builder()
+                                .nOut(40)
+                                .activation(Activation.TANH)
+                                .build(), "input")
+                        .addLayer("corridor", new DenseLayer.Builder().nOut(40).activation(Activation.RELU).build(), "input", "goal")
+                        .addLayer("corridor-1", new DenseLayer.Builder().nOut(20).activation(Activation.RELU).build(), "corridor")
+                        .addVertex("corridor-rnn", new PreprocessorVertex(new FeedForwardToRnnPreProcessor()), "corridor-1");
+    }

```

Comment:
Does this mean users will need to build neural networks on their own? There won't be any defaults?

Reply:
Users will indeed need to build their own neural nets. Since every environment is different, a user will need to use his expertise and intuition the same way he would if he created a network for a new deep learning task. I think DL4J has tools to help create networks? Like adding a 'convolutional block' to a network? These will be useful. And also, I have a (long term) objective to be able to use Arbiter with RL4j, to help fine-tune the network's hyperparameters but also the reinforcement learning hyperparameters (gamma, training batch size, ...)

We could however offer defaults networks for our environments 'wrappers' (GYM, ALE, Malmo, ...)

Reply:
Users shouldn't need to define custom networks, no. Of course, we need to be able to add new networks, but it shouldn't be required. They should also not be associated with the environments. They need to be independent, well, actually part of the configuration of the RL algorithms, not of the environments.

Reply:
It's not that simple unfortunately. As you know, some environments require RNNs, other convolutions, others will do just fine with dense layers, etc... And the type of each layer and how they are connected together will have an enormous impact on how well they train. For example, this is the network that solves the TMaze environment:

![image](https://user-images.githubusercontent.com/44292157/91915010-72dc4600-ec87-11ea-8453-28ef6daa09ba.png)

This environment require the network to remember the first observation until the end of the maze. This is why the LSTM is connected to the output.
Now, if we use instead CartPole and we remove the angular momentum from the state.  The network will also need a LSTM to approximate the missing information, but if we use the same topology as for TMaze, it will train very slowly (or not at all) because the direct link between the LSTM and the output is detrimental. If we just remove that link, like this:

![image](https://user-images.githubusercontent.com/44292157/91915062-943d3200-ec87-11ea-9a59-f1ae6ba10b3c.png)

The new network learns much faster. 

Don't get me wrong, having default networks would be very nice. But I havent't found how we should do it.
Do you have an idea?

Reply:
I'm not saying it shouldn't support custom networks, I'm saying there should be reasonable defaults for each algorithm.

Reply:
I understand and agree. I think we should add several 'default' (or more realistically, a few builders that can be configured into several networks). A user will be able to test several different topologies and either pick one that works well or at least have a good starting point to design his own. But if you don't mind it to be done later, I'd like to finish the recurrent support, then add the training logging support, and add load/save support before the defaults.

Reply:
Ok, but how are the defaults going to work? It pretty much needs to be in some configuration for each algorithm somewhere. For now, let's at least decide how that's going to look like?

Reply:
Ok, here are my first thoughts:
The first thing is to move the network to the configuration. The configuration should support being passed either directly a network instance, or an instance of `Builder<ITrainableNeuralNet>`.

If a `Builder<ITrainableNeuralNet>` is used, and since the output of a network is determined by the algorithm used, we could be nice and offer an option to automatically add the correct outputs to the network.

Instead of just offering default network, we could also offer a way to build simple networks. With this simple network builder, simple blocks are stacked such as:
- DenseLayerBlock: a block of N `DenseLayer`, with an optional max-pool at the end.
- CNNBlock: a block of N `ConvolutionLayer`, also with an optional max-pool at the end.
- RNNBlock: a block of N `LSTM` layers

Within a block, the size of individual layers can be annealed to create a funnel-like block (e.g. 20-10-5), a square block (e.g. 20-20-20) or a inverse-funnel (e.g. 5-10-20)
A block is always connected to the output of the preceding block, and optionally directly connected to the network output layer(s) (except the last block which is always connected to the network output, obviously).

Finally, we can offer a few default networks, which will internally use the above builder, such as:
- A default network for image based environments
- A default network for memory-based networks or POMDPs (a recurrent net.)
- A default network for fully observable state networks (dense layers only)

What do you think?

Reply:
Yes, something like that should work. There should still be defaults like there are for all other configuration parameters though.